Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMWidevineDecryptorGStreamer.cpp

```diff
@@ -0,0 +1,226 @@
+            return false;
+        }
+
+        GUniquePtr<GstByteReader> reader(gst_byte_reader_new(subSamplesMap.data, subSamplesMap.size));
+        uint16_t inClear = 0;
+        uint32_t inEncrypted = 0;
+        uint32_t totalEncrypted = 0;
+        unsigned position;
+        // Find out the total size of the encrypted data.
+        for (position = 0; position < subSampleCount; position++) {
+            gst_byte_reader_get_uint16_be(reader.get(), &inClear);
+            gst_byte_reader_get_uint32_be(reader.get(), &inEncrypted);
+            totalEncrypted += inEncrypted;
+        }
+        gst_byte_reader_set_pos(reader.get(), 0);
+
+        // Build a new buffer storing the entire encrypted cipher.
+        GUniquePtr<uint8_t> holdEncryptedData(reinterpret_cast<uint8_t*>(malloc(totalEncrypted)));
+        uint8_t* encryptedData = holdEncryptedData.get();
+        unsigned index = 0;
+        for (position = 0; position < subSampleCount; position++) {
+            gst_byte_reader_get_uint16_be(reader.get(), &inClear);
+            gst_byte_reader_get_uint32_be(reader.get(), &inEncrypted);
+            memcpy(encryptedData, map.data + index + inClear, inEncrypted);
+            index += inClear + inEncrypted;
+            encryptedData += inEncrypted;
+        }
+        gst_byte_reader_set_pos(reader.get(), 0);
+
+        // Decrypt cipher.
+        int errorCode;
+        if (errorCode = priv->m_openCdm->Decrypt(holdEncryptedData.get(), static_cast<uint32_t>(totalEncrypted),
+            ivMap.data, static_cast<uint32_t>(ivMap.size))) {
+            GST_WARNING_OBJECT(self, "ERROR - packet decryption failed [%d]", errorCode);
+            gst_buffer_unmap(buffer, &map);
+            gst_buffer_unmap(subSamplesBuffer, &subSamplesMap);
+            gst_buffer_unmap(ivBuffer, &ivMap);
+            return false;
+        }
+
+        // Re-build sub-sample data.
+        index = 0;
+        encryptedData = holdEncryptedData.get();
+        unsigned total = 0;
+        for (position = 0; position < subSampleCount; position++) {
+            gst_byte_reader_get_uint16_be(reader.get(), &inClear);
+            gst_byte_reader_get_uint32_be(reader.get(), &inEncrypted);
+
+            memcpy(map.data + total + inClear, encryptedData + index, inEncrypted);
+            index += inEncrypted;
+            total += inClear + inEncrypted;
+        }
+
+        gst_buffer_unmap(subSamplesBuffer, &subSamplesMap);
+    } else {
+        // Decrypt cipher.
+        int errorCode;
+        if (errorCode = priv->m_openCdm->Decrypt(map.data, static_cast<uint32_t>(map.size),
+            ivMap.data, static_cast<uint32_t>(ivMap.size))) {
+            GST_WARNING_OBJECT(self, "ERROR - packet decryption failed [%d]", errorCode);
+            gst_buffer_unmap(buffer, &map);
+            gst_buffer_unmap(ivBuffer, &ivMap);
+            return false;
+        }
+    }
+    gst_buffer_unmap(buffer, &map);
+    gst_buffer_unmap(ivBuffer, &ivMap);
```

Comment:
The unmaps is good code. I don't think we are leaving anything behind when we are doing error control. The problem is that we are repeating the unmap every time we are doing error check.

I think we can use here the goto antipattern but that is a common pattern in GStreamer code:

  int returnValue = true;

...

  } else {
      // Decrypt cipher.
      int errorCode;
      if (errorCode = priv->m_openCdm->Decrypt(map.data, static_cast<uint32_t>(map.size),
          ivMap.data, static_cast<uint32_t>(ivMap.size))) {
          GST_WARNING_OBJECT(self, "ERROR - packet decryption failed [%d]", errorCode);
          returnValue = false;
          goto beach;
      }
    }

 beach:
  gst_buffer_unmap(buffer, &map);
  gst_buffer_unmap(ivBuffer, &ivMap);

Of course you would assign returnValue to false and goto beach in more places than the one I stated. For subSamplesMap the unmap is done only twice, I don't think there is need to move that to the beach, just unmap the normal and the error.